library/std/src/io/mod.rs

```diff
@@ -386,20 +397,12 @@ where
         }
 
         match r.read(&mut g.buf[g.len..]) {
-            Ok(0) => {
-                ret = Ok(g.len - start_len);
-                break;
-            }
+            Ok(0) => return Ok(g.len - start_len),
             Ok(n) => g.len += n,
```

Comment:
Could we check for an overflow here?

Reply:
We could, but I would rather keep the codepath oriented as much as possible towards non-malicious implementations. Overflow can only be caused by bogus return values, and just means that a nonsensical result will be returned.

Reply:
Otherwise, malicious `Read` can shrink `g.len`, which can lead to a violation of `append_to_string()`'s second requirement. (and `append_to_string()` should probably be marked as an unsafe function)

https://github.com/rust-lang/rust/blob/c97f11af7bc4a6d3578f6a953be04ab2449a5728/library/std/src/io/mod.rs#L320-L329

Reply:
This is the context:
https://github.com/rust-lang/rust/blob/c97f11af7bc4a6d3578f6a953be04ab2449a5728/library/std/src/io/mod.rs#L707-L718

Reply:
Ah right. We'll need to replace that with a checked_add then.

This kind of mess is one of the things that the ReadBuf proposal is designed to avoid, so things should get less janky when that's implemented.